emily/handler/src/database/entries/mod.rs

```diff
@@ -1,8 +1,435 @@
+//!   - `SecondaryIndex<T>`: Wrapper for secondary index structures.
+//!
+//! - **Traits**:
+//!   - `EntryTrait`: A trait common to all database entries, defining how entries are keyed.
+//!   - `KeyTrait`: Defines the structure of a table key, including partition and sort keys.
+//!   - `PrimaryIndexTrait`: Used for defining a primary index and its associated table operations.
+//!   - `SecondaryIndexTrait`: Extends the `PrimaryIndexTrait` to support secondary indexes.
+//!   - `TableIndexTrait`: A base trait that defines common table operations across indexes.
+//!
+//! - **Implementations**:
+//!   - Generic implementations for handling database operations such as get, query, put, and delete
+//!     across different types of indexes.
+//!
+//! ## Database Operations
+//!
+//! This module abstracts the complexity of interacting with DynamoDB by providing easy-to-use
+//! methods for common operations such as:
+//!
+//! - **Retrieval**: Fetch entries by key or query based on a partition key.
+//! - **Insertion**: Add new entries to the database.
+//! - **Deletion**: Remove entries from the database either by key or in bulk.
+//! - **Scanning**: Retrieve all entries from a table.
+//! - **Table Management**: Wipe a table clean for testing or other purposes.
+//!
+//! ## Usage
+//!
+//! This module is intended for internal use within the application and is designed to be flexible
+//! enough to support various data models while maintaining strong type safety and clear separation
+//! of concerns.
+
+use std::collections::HashMap;
+
+use aws_sdk_dynamodb::types::{AttributeValue, DeleteRequest, WriteRequest};
+use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine};
+use serde::{Deserialize, Serialize};
+use serde_dynamo::Item;
+
+use crate::{common::error::Error, context::Settings};
 
 /// Chainstate table entries.
 pub mod chainstate;
 /// Deposit table entries.
 pub mod deposit;
 /// Withdrawal table entries.
 pub mod withdrawal;
+
+// Structures
+// -----------------------------------------------------------------------------
+
+/// Primary index wrapper struct.
+pub struct PrimaryIndex<T>(pub T);
+
+/// Secondary index wrapper struct.
+pub struct SecondaryIndex<T>(pub T);
+
+// Traits
+// -----------------------------------------------------------------------------
+
+/// Trait common to entries.
+pub trait EntryTrait: serde::Serialize + for<'de> serde::Deserialize<'de> {
+    /// Key type for the entry.
+    type Key: KeyTrait;
+    /// Retrieves the entry key.
+    fn key(&self) -> Self::Key;
+}
+
+/// Table key.
+pub trait KeyTrait: serde::Serialize + for<'de> serde::Deserialize<'de> {
+    /// Partition key type.
+    type PartitionKey: serde::Serialize + for<'de> serde::Deserialize<'de>;
+    /// Sort key type.
+    type SortKey: serde::Serialize + for<'de> serde::Deserialize<'de>;
+    /// Parition key name.
+    const PARTITION_KEY_NAME: &'static str;
+    /// Sort key name.
+    const _SORT_KEY_NAME: &'static str;
```

Comment:
Yeah, it's not used anywhere but I'd rather keep it in for completion since these are all the parameters of the index. If we need it later we would need to refer to the table creation and add it back in - the underscore removes the confusion about it being unused while still allowing us to specify. IMO that's better than not having it.

Reply:
So you can see, these are the features specified when creating the secondary index: https://github.com/stacks-network/sbtc/blob/main/emily/cdk/lib/emily-stack.ts#L135-L145